hail/python/hail/methods/statgen.py

```diff
@@ -444,43 +437,42 @@ def select_array_indices(hl_array, indices):
     def dot_rows_with_themselves(matrix):
         return (matrix * matrix) @ hl.nd.ones(matrix.shape[1])
 
-    def array_from_struct(struct, field_names):
-        return hl.array([struct[field_name] for field_name in field_names])
+    def no_missing(hail_array):
+        return hail_array.all(lambda element: hl.is_defined(element))
 
     ht_local = mt._localize_entries(entries_field_name, sample_field_name)
 
     ht = ht_local.transmute(**{entries_field_name: ht_local[entries_field_name][x_field_name]})
 
-    def make_one_cov_matrix(ys_and_covs_to_keep):
-        return hl.nd.array(ys_and_covs_to_keep.map(lambda struct: array_from_struct(struct, cov_field_names))) \
-            if cov_field_names else hl.nd.zeros((hl.len(ys_and_covs_to_keep), 0))
-
-    def make_one_y_matrix(ys_and_covs_to_keep, one_y_field_name_set):
-        return hl.nd.array(ys_and_covs_to_keep.map(lambda struct: array_from_struct(struct, one_y_field_name_set)))
-
     def setup_globals(ht):
-        all_covs_defined = ht[sample_field_name].map(lambda sample_struct: all_defined(sample_struct, cov_field_names))
-        list_of_ys_and_covs_to_keep_with_indices = [
-            hl.enumerate(ht[sample_field_name]).filter(
-                lambda struct_with_index:
-                all_covs_defined[struct_with_index[0]] & all_defined(struct_with_index[1], one_y_field_name_set))
-            for one_y_field_name_set in y_field_names
-        ]
+        # cov_arrays is per sample, then per cov.
+        if covariates:
+            ht = ht.annotate_globals(cov_arrays=ht[sample_field_name].map(lambda sample_struct: [sample_struct[cov_name] for cov_name in cov_field_names]))
+        else:
+            ht = ht.annotate_globals(cov_arrays=ht[sample_field_name].map(lambda sample_struct: hl.empty_array(hl.tfloat64)))
 
-        list_of_ys_and_covs_to_keep = [inner_list.map(lambda pair: pair[1]) for inner_list in list_of_ys_and_covs_to_keep_with_indices]
-        kept_samples = [inner_list.map(lambda pair: pair[0]) for inner_list in list_of_ys_and_covs_to_keep_with_indices]
+        ht = ht.annotate_globals(
+            y_arrays_per_group=[ht[sample_field_name].map(lambda sample_struct: [sample_struct[y_name] for y_name in one_y_field_name_set]) for one_y_field_name_set in y_field_name_groups]
```

Comment:
I think the ideal code would be to make an array of arrays of the field indices for each group, then index into the sample structs in a map. But I don't think we support dynamic tuple indices. Might be worth considering how hard that would be, for future performance tuning.

Reply:
It's true, we don't support dynamic tuple indices.